File: Daily-Streak/Day-07_Maximum-Product-of-Splitted-Binary-Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private long totalSum = 0;
    private long maxProduct = 0;
    private final int MOD = 1000000007;

    // First DFS: total tree sum
    private long getTotalSum(TreeNode root) {
        if (root == null) return 0;
        return root.val + getTotalSum(root.left) + getTotalSum(root.right);
    }

    // Second DFS: subtree sums & max product
    private long dfs(TreeNode root) {
        if (root == null) return 0;

        long leftSum = dfs(root.left);
        long rightSum = dfs(root.right);

        long subtreeSum = root.val + leftSum + rightSum;

        long product = subtreeSum * (totalSum - subtreeSum);
        maxProduct = Math.max(maxProduct, product);

        return subtreeSum;
    }

    public int maxProduct(TreeNode root) {
        // ✅ OPTIMAL TWO-PASS DFS APPROACH (ACTIVE)
        totalSum = getTotalSum(root);
        dfs(root);
        return (int)(maxProduct % MOD);
    }

    /*
    ===========================
    ❌ NAIVE EDGE-REMOVAL IDEA (COMMENTED)
    ===========================
    - Remove edge
    - Compute sums repeatedly
    - Very slow

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        Solution s = new Solution();
        System.out.println(s.maxProduct(root));
    }
    */
}

File: Daily-Streak/Day-07_Maximum-Product-of-Splitted-Binary-Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long totalSum = 0;
    long long maxProd = 0;
    const int MOD = 1e9 + 7;

    // First DFS: compute total sum of tree
    long long getTotalSum(TreeNode* root) {
        if (!root) return 0;
        return root->val + getTotalSum(root->left) + getTotalSum(root->right);
    }

    // Second DFS: compute subtree sums & update max product
    long long dfs(TreeNode* root) {
        if (!root) return 0;

        long long leftSum = dfs(root->left);
        long long rightSum = dfs(root->right);

        long long subtreeSum = root->val + leftSum + rightSum;

        // Try splitting here
        long long product = subtreeSum * (totalSum - subtreeSum);
        maxProd = max(maxProd, product);

        return subtreeSum;
    }

    int maxProduct(TreeNode* root) {
        // ✅ OPTIMAL TWO-PASS DFS APPROACH (ACTIVE)
        totalSum = getTotalSum(root);
        dfs(root);
        return maxProd % MOD;
    }

    /*
    ===========================
    ❌ BRUTE FORCE IDEA (COMMENTED)
    ===========================
    - Remove each edge
    - Recompute subtree sums
    - O(n^2) → TLE

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    #include <bits/stdc++.h>
    using namespace std;

    int main() {
        TreeNode* root = new TreeNode(1);
        root->left = new TreeNode(2);
        root->right = new TreeNode(3);
        root->left->left = new TreeNode(4);
        root->left->right = new TreeNode(5);

        Solution s;
        cout << s.maxProduct(root) << endl;
        return 0;
    }
    */
};
